firewood/src/merkle.rs

```diff
@@ -1465,58 +1341,81 @@ mod test {
-            let new_encoded = new_node.get_encoded(merkle.store.as_ref());
-            assert_eq!(r, new_encoded);
+    fn branch(value: Vec<u8>, encoded_child: Option<Vec<u8>>) -> Node {
+        let children = Default::default();
+        let value = Some(value).map(Data);
+        let mut children_encoded = <[Option<Vec<u8>>; NBRANCH]>::default();
+
+        if let Some(child) = encoded_child {
+            children_encoded[0] = Some(child);
         }
 
-        {
-            let chd = Node::new(NodeType::Branch(BranchNode {
-                children: [None; NBRANCH],
-                value: Some(Data("value1".as_bytes().to_vec())),
-                children_encoded: Default::default(),
-            }));
-            let chd_ref = merkle.put_node(chd.clone()).unwrap();
-            let chd_encoded = chd_ref.get_encoded(merkle.store.as_ref());
-            let new_chd = Node::new(NodeType::decode(chd_encoded).unwrap());
-            let new_chd_encoded = new_chd.get_encoded(merkle.store.as_ref());
-            assert_eq!(chd_encoded, new_chd_encoded);
-
-            let node = Node::new(NodeType::Extension(ExtNode {
-                path: PartialPath(vec![0x1, 0x2, 0x3]),
-                child: DiskAddress::null(),
-                child_encoded: Some(chd_encoded.to_vec()),
-            }));
-            let node_ref = merkle.put_node(node.clone()).unwrap();
-
-            let r = node_ref.get_encoded(merkle.store.as_ref());
-            let new_node = Node::new(NodeType::decode(r).unwrap());
-            let new_encoded = new_node.get_encoded(merkle.store.as_ref());
-            assert_eq!(r, new_encoded);
+        Node::branch(BranchNode {
+            children,
+            value,
+            children_encoded,
+        })
+    }
+
+    #[test_case(leaf(vec![1, 2, 3], vec![4, 5]) ; "leaf encoding")]
+    #[test_case(branch(b"value".to_vec(), vec![1, 2, 3].into()) ; "branch with value")]
+    #[test_case(branch(b"value".to_vec(), None); "branch without value")]
+    #[test_case(extension(vec![1, 2, 3], DiskAddress::null(), vec![4, 5].into()) ; "extension without child address")]
+    fn encode_(node: Node) {
+        let merkle = create_test_merkle();
+
+        let node_ref = merkle.put_node(node).unwrap();
+        let encoded = node_ref.get_encoded(merkle.store.as_ref());
+        let new_node = Node::from(NodeType::decode(encoded).unwrap());
+        let new_node_encoded = new_node.get_encoded(merkle.store.as_ref());
+
+        assert_eq!(encoded, new_node_encoded);
+    }
+
+    #[test]
+    fn insert_and_retrieve() {
+        let key = b"hello";
+        let val = b"world";
+
+        let mut merkle = create_test_merkle();
+        let root = merkle.init_root().unwrap();
+
+        merkle.insert(key, val.to_vec(), root).unwrap();
+
+        let fetched_val = merkle.get(key, root).unwrap();
+
+        assert_eq!(fetched_val.as_deref(), val.as_slice().into());
+    }
+
+    #[test]
+    fn insert_and_retrieve_multiple() {
```

Comment:
I guess I should explain why this is necessary, what do you think?

Reply:
🤔, you're absolutely right, however, by the same argument, one could say that many unit tests are actually covered by integration tests and are therefore unnecessary. 

There are three separate outcomes when looking at these tests together:
1. both tests pass (everything works, hurray)
2. the `_multiple` test fails and the other tests pass (something is wrong with multiple insertion specifically, it's probably something to do with splitting nodes)
3. both tests fail (nothing is working)

And with that said, I've realized that this test needs to be altered as it would not cause a split, but I am going to do that in the next PR!